pkg/reconciler/namespace/namespace.go

```diff
@@ -0,0 +1,202 @@
+/*
+Copyright 2019 The Knative Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package namespace
+
+import (
+	"bytes"
+	"context"
+	"strings"
+	"text/template"
+
+	"github.com/knative/pkg/controller"
+	"github.com/knative/pkg/logging"
+	"github.com/knative/serving/pkg/apis/networking/v1alpha1"
+	listers "github.com/knative/serving/pkg/client/listers/networking/v1alpha1"
+	"github.com/knative/serving/pkg/network"
+	"github.com/knative/serving/pkg/reconciler"
+	"github.com/knative/serving/pkg/reconciler/namespace/config"
+	"github.com/knative/serving/pkg/reconciler/namespace/resources/names"
+	"go.uber.org/zap"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/equality"
+	apierrs "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	kubelisters "k8s.io/client-go/listers/core/v1"
+	"k8s.io/client-go/tools/cache"
+)
+
+const (
+	blacklistLabel = "networking.internal.knative.dev/namespaceCert"
+)
+
+var wildcardName = names.WildcardCertName
+
+// Reconciler implements controller.Reconciler for Certificate resources.
+type Reconciler struct {
+	*reconciler.Base
+
+	// listers index properties about resources
+	nsLister            kubelisters.NamespaceLister
+	knCertificateLister listers.CertificateLister
+
+	configStore configStore
+}
+
+// Check that our Reconciler implements controller.Reconciler
+var _ controller.Reconciler = (*Reconciler)(nil)
+
+func (c *Reconciler) Reconcile(ctx context.Context, key string) error {
+	// Convert the namespace/name string into a distinct namespace and name
+	_, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		c.Logger.Errorf("invalid resource key: %s", key)
+		return nil
+	}
+	logger := logging.FromContext(ctx)
+	ctx = c.configStore.ToContext(ctx)
+
+	original, err := c.nsLister.Get(name)
+
+	if apierrs.IsNotFound(err) {
+		logger.Errorf("Namespace %s in work queue no longer exists %s", key, err)
+		return nil
+	} else if err != nil {
+		return err
+	}
+
+	// Don't modify the informers copy
+	newNamespace := original.DeepCopy()
+
+	allocateCert := false
+	for label, value := range newNamespace.Labels {
+		if label == blacklistLabel && value == "true" {
```

Comment:
this looks more like a whitelist, than a blacklist

Reply:
It is. I'll change the names.